Starcraft2Client.py

```diff
@@ -568,169 +561,116 @@ async def on_step(self, iteration: int):
+        message += ", ".join(f"{mark_up_mission_name(ctx, mission, unlocks)}[{ctx.mission_req_table[mission].id}] " +
                              mark_up_objectives(
-                                 f"[{unfinished_missions[mission]}/{location_table[mission].extra_locations}]",
+                                 f"[{unfinished_missions[mission]}/"
+                                 f"{sum(1 for _ in ctx.locations_for_mission(mission))}]",
                                  ctx, unfinished_locations, mission)
                              for mission in unfinished_missions)
 
-        if ui:
-            ui.log_panels['All'].on_message_markup(message)
-            ui.log_panels['Starcraft2'].on_message_markup(message)
+        if ctx.ui:
+            ctx.ui.log_panels['All'].on_message_markup(message)
+            ctx.ui.log_panels['Starcraft2'].on_message_markup(message)
         else:
             sc2_logger.info(message)
     else:
         sc2_logger.warning("No mission table found, you are likely not connected to a server.")
 
 
-def calc_unfinished_missions(locations_done, locations, ctx, unlocks=None, unfinished_locations=None,
-                             available_missions=[]):
+def calc_unfinished_missions(ctx: SC2Context, unlocks=None):
     unfinished_missions = []
     locations_completed = []
 
     if not unlocks:
-        unlocks = initialize_blank_mission_dict(locations)
-
-    if not unfinished_locations:
-        unfinished_locations = initialize_blank_mission_dict(locations)
-
-    if len(available_missions) > 0:
-        available_missions = []
+        unlocks = initialize_blank_mission_dict(ctx.mission_req_table)
 
-    available_missions.extend(calc_available_missions(locations_done, locations, unlocks))
+    available_missions = calc_available_missions(ctx, unlocks)
 
     for name in available_missions:
-        if not locations[name].extra_locations == -1:
-            objectives_completed = calc_objectives_completed(name, locations, locations_done, unfinished_locations, ctx)
+        objectives = tuple(ctx.locations_for_mission(name))
+        if objectives:
+            objectives_completed = calc_objectives_completed(ctx, name)
 
-            if objectives_completed < locations[name].extra_locations:
+            if objectives_completed < len(objectives):
                 unfinished_missions.append(name)
                 locations_completed.append(objectives_completed)
 
-        else:
+        else:  # infer that this is the final mission as it has no objectives
             unfinished_missions.append(name)
             locations_completed.append(-1)
 
-    return {unfinished_missions[i]: locations_completed[i] for i in range(len(unfinished_missions))}
+    return available_missions, {unfinished_missions[i]: locations_completed[i]
+                                for i in range(len(unfinished_missions))}
```

Comment:
this looks like there should be an `enumerate` somewhere

Reply:
enumerate would not work as it's two merged lists, changed over to dict(zip()) though

Reply:
my first idea was `mission: locations_completed[i] for i, mission in enumerate(unfinished_missions)`, but yeah, zip is even better